contracts/modules/cover/CoverOrder.sol

```diff
@@ -0,0 +1,225 @@
+    address _memberRoles,
+    address _nxmToken,
+    address _master,
+    address _weth,
+    address _owner
+  ) {
+    master = INXMMaster(_master);
+    cover = ICover(_cover);
+    memberRoles = IMemberRoles(_memberRoles);
+    nxmToken = INXMToken(_nxmToken);
+    weth = IWeth(_weth);
+    transferOwnership(_owner);
+  }
+
+  /// @notice Verifies and executes the order to buy cover on behalf of the creator of limit order
+  /// @param params Cover buy parameters
+  /// @param poolAllocationRequests Pool allocations for the cover
+  /// @param limitOrderId The ID of the limit order
+  /// @param signature The signature of the order
+  function executeOrder(
+    BuyCoverParams calldata params,
+    PoolAllocationRequest[] calldata poolAllocationRequests,
+    uint limitOrderId,
+    bytes calldata signature
+  ) external payable returns (uint coverId) {
+
+    if (params.owner == address(0) || params.owner == address(this)) {
+      revert InvalidOwnerAddress();
+    }
+
+    if (params.paymentAsset == NXM_ASSET_ID) {
+      revert InvalidPaymentAsset();
+    }
+
+    // Compute hash
+    bytes32 orderHash = keccak256(
+      abi.encode(
+        params.productId,
+        params.coverId,
+        params.amount,
+        params.period,
+        params.paymentAsset,
+        params.coverAsset,
+        params.owner,
+        limitOrderId
+      )
+    );
+
+    // Ensure the order has not already been executed
+    if (executedOrders[orderHash]) {
+      revert OrderAlreadyExecuted(limitOrderId);
+    }
+
+    // Recover the signer address from the signature
+    bytes32 ethSignedOrderHash = orderHash.toEthSignedMessageHash();
+    address signer = ethSignedOrderHash.recover(signature);
+
+    // Ensure the signer is the user who owns the order
+    if (signer != params.owner) {
+      revert InvalidSignature();
+    }
+
+    // Mark the order as executed
+    executedOrders[orderHash] = true;
+
+    // ETH payment
+    if (params.paymentAsset == ETH_ASSET_ID) {
+      return _buyCoverEthPayment(params, poolAllocationRequests);
+    }
+
+    // ERC20 payment
+    return _buyCoverErc20Payment(params, poolAllocationRequests);
+  }
+
+  /// @notice Handles ETH/WETH payments for buying cover.
+  /// @dev Transfers WETH tokens from the order creator to the contract, then unwraps it,  then buys cover on behalf of the creator.
+  //       Calculates ETH refunds if any and sends back to param.owner.
+  /// @param params The parameters required to buy cover.
+  /// @param poolAllocationRequests The allocation requests for the pool's liquidity.
+  /// @return coverId The ID of the purchased cover.
+  function _buyCoverEthPayment(
+    BuyCoverParams calldata params,
+    PoolAllocationRequest[] calldata poolAllocationRequests
+  ) internal returns (uint coverId) {
+
+    uint ethBalanceBefore = address(this).balance;
+
+    weth.transferFrom(params.owner, address(this), params.maxPremiumInAsset);
+    weth.withdraw(params.maxPremiumInAsset);
```

Comment:
why do the payment via WETH? why not just do ETH payment?

Reply:
There is no option to allow contract to transfer ETH from your wallet. It is only possible to do it with ERC20 tokens.

Reply:
Other way would be to transfer ETH beforehand which I wouln't do

Reply:
i'm with you now, we only want to pull in the moneys when we hit the our target price not before.